src/helpers.coffee

```diff
@@ -111,30 +111,43 @@ buildLocationData = (first, last) ->
 buildLocationHash = (loc) ->
   "#{loc.first_line}x#{loc.first_column}-#{loc.last_line}x#{loc.last_column}"
 
+# Build a dictionary of extra token properties organized by tokens’ locations
+# used as lookup hashes.
+buildTokenDataDictionary = (parserState) ->
+  tokenData = {}
+  for token in parserState.parser.tokens when token.comments
+    tokenHash = buildLocationHash token[2]
+    # Multiple tokens might have the same location hash, such as the generated
+    # `JS` tokens added at the start or end of the token stream to hold
+    # comments that start or end a file.
+    tokenData[tokenHash] ?= {}
+    if token.comments # `comments` is always an array.
+      # For “overlapping” tokens, that is tokens with the same location data
+      # and therefore matching `tokenHash`es, merge the comments from both/all
+      # tokens together into one array, even if there are duplicate comments;
+      # they will get sorted out later.
+      if tokenData[tokenHash].comments?
```

Comment:
@GeoffreyBooth I know this is just basically moving existing code, but a way I like to write this pattern is:
```
(tokenData[tokenHash].comments ?= []).push token.comments...
```